R/bm-tpc-h.R

```diff
@@ -218,20 +232,397 @@ tpc_h_queries[[1]] <- function(input_func) {
+    p, by = c("ps_partkey" = "p_partkey"))
+
+  sn <- inner_join(
+    input_func("supplier") %>%
+      select(s_suppkey, s_nationkey),
+    input_func("nation") %>%
+      select(n_nationkey, n_name),
+    by = c("s_nationkey" = "n_nationkey")) %>%
+    select(s_suppkey, n_name)
+
+  pspsn <- inner_join(psp, sn, by = c("ps_suppkey" = "s_suppkey"))
+
+  lpspsn <- inner_join(
+    input_func("lineitem") %>%
+      select(l_suppkey, l_partkey, l_orderkey, l_extendedprice, l_discount, l_quantity),
+    pspsn,
+    by = c("l_suppkey" = "ps_suppkey", "l_partkey" = "ps_partkey")) %>%
+    select(l_orderkey, l_extendedprice, l_discount, l_quantity, ps_supplycost, n_name)
+
+  all <- inner_join(
+    input_func("orders") %>%
+      select(o_orderkey, o_orderdate),
+    lpspsn,
+    by = c("o_orderkey"= "l_orderkey" )) %>%
+    select(l_extendedprice, l_discount, l_quantity, ps_supplycost, n_name, o_orderdate)
+
+  aggr <- all %>%
+    # kludge, should not need to collect here, but if we don't the answers we
+    # get are wrong (though they *do* complete).
+    collect() %>%
+    mutate(
+      nation = n_name,
+      # kludge, o_year = as.integer(format(o_orderdate, "%Y")),
+      o_year = as.integer(strftime(o_orderdate, "%Y")),
+      amount = l_extendedprice * (1 - l_discount) - ps_supplycost * l_quantity) %>%
+    select(nation, o_year, amount) %>%
+    group_by(nation, o_year) %>%
+    summarise(sum_profit = sum(amount)) %>%
+    arrange(nation, desc(o_year))
+
+  aggr %>%
+    collect()
+}
+
+tpc_h_queries[[10]] <- function(input_func) {
+  l <- input_func("lineitem") %>%
+    select(l_orderkey, l_returnflag, l_extendedprice, l_discount) %>%
+    filter(l_returnflag == "R") %>%
+    select(l_orderkey, l_extendedprice, l_discount)
+
+  o <- input_func("orders") %>%
+    select(o_orderkey, o_custkey, o_orderdate) %>%
+    # kludge, filter(o_orderdate >= "1993-10-01", o_orderdate < "1994-01-01") %>%
+    filter(o_orderdate >= as.Date("1993-10-01"), o_orderdate < as.Date("1994-01-01")) %>%
+    select(o_orderkey, o_custkey)
+
+  lo <- inner_join(l, o,
+                   by = c("l_orderkey" = "o_orderkey")) %>%
+    select(l_extendedprice, l_discount, o_custkey)
+  # first aggregate, then join with customer/nation,
+  # otherwise we need to aggr over lots of cols
+
+  lo_aggr <- lo %>% mutate(volume=l_extendedprice * (1 - l_discount)) %>%
+    group_by(o_custkey) %>%
+    summarise(revenue = sum(volume))
+
+  c <- input_func("customer") %>%
+    select(c_custkey, c_nationkey, c_name, c_acctbal, c_phone, c_address, c_comment)
+
+  loc <- inner_join(lo_aggr, c, by = c("o_custkey" = "c_custkey"))
+
+  locn <- inner_join(loc, input_func("nation") %>% select(n_nationkey, n_name),
+                     by = c("c_nationkey" = "n_nationkey"))
+
+  res <- locn %>%
+    select(o_custkey, c_name, revenue, c_acctbal, n_name,
+           c_address, c_phone, c_comment) %>%
+    # kludge, should not need to collect here
```

Comment:
link to the jira where we'll fix this? also `compute()` instead?

Reply:
I added the link. I've also added to the other places we use this pattern to do `arrange() %>% compute() %>% head() %>% collect()` which is more characters, but should be a bit faster since we're not collecting everything into R